gh-91400: make sure email parsing dont unquote realnames with spaces

carlbordum opened the pull request:
"Quote the name" portion of an email address in `email.utils.formataddr()` when it contains any spaces.

<!-- gh-issue-number: gh-91400 -->
* Issue: gh-91400
<!-- /gh-issue-number -->

carlbordum committed:
gh-91400: make sure email parsing dont unquote realnames with spaces

Akuli commented on Lib/test/test_email/test_email.py:
```diff
@@ -3164,6 +3164,14 @@ def test_escape_backslashes(self):
+        # gh-91400
+        identity = '"foo  bar" <foo@example.com>'
+        single = utils.formataddr(utils.parseaddr(identity))
+        double = utils.formataddr(utils.parseaddr(single))
```
Do we really need to repeat the check twice? It doesn't make sense to me. If the first line returns the string unchanged, we have `identity == single` anyway, and so the second line gets the same string as the first line.

carlbordum replied:
Before the change in `email/utils.py`, this unit tests fails. Although your comment reflects what would be expected. See the issue (https://github.com/python/cpython/issues/91400) for example.

Akuli replied:
My point is that if it doesn't change on the first try, it won't change on the second try either. The second call is just unnecessary and confusing. In the issue it changes on the first call, and that's what the test should check.

gpshead committed:
Restrict NEWS wording to the exact behavior change.

This way it describes the observable behavior change. Rather than the more broad wording that'd make a guarantee we don't actually test exhaustively for.

gpshead committed:
simplify the test

running it a second time when a round tripped identical value has already been guaranteed doesn't do anything.  Though it was meaningful during the original bug report, in correct code it is not.

gpshead committed:
Add a comment describing the purpose of the regexes.

arhadthedev committed:
Merge branch 'main' into 91400-formataddr

hugovk committed:
Merge branch 'main' into 91400-formataddr

hugovk reviewed (approved):
> any objections from folks on email-team?

~Two years later and no objections, let's merge this.

bitdancer reviewed (changes requested):
formataddr does the right thing.  It's parseaddr that would need a fix, as well as _headervalueparser.
